# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: caps/nsScriptSecurityManager.cpp
# Commit: 955256297d6d
# Full Hash: 955256297d6dc83d52eb77a202c9c320622c0d13
# Author: Jan de Mooij <jdemooij@mozilla.com>
# Date: 2019-11-05 16:16:12
# Regressor Bug: 1579367
# File Overlap Count: 1
# Description:
#   Bug 1579367 - Initialize XPCJSContext explicitly, after loading user prefs. r=kmag
#   
#   This way we get the correct values for start-up prefs in the parent process.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D51061
# ==============================================================================

diff -r 68e7fb2885b0 -r 955256297d6d caps/nsScriptSecurityManager.cpp
--- a/caps/nsScriptSecurityManager.cpp	Tue Nov 05 08:19:10 2019 +0200
+++ b/caps/nsScriptSecurityManager.cpp	Tue Nov 05 06:29:13 2019 +0000
@@ -85,7 +85,6 @@
 using namespace mozilla::dom;
 
 nsIIOService* nsScriptSecurityManager::sIOService = nullptr;
-JSContext* nsScriptSecurityManager::sContext = nullptr;
 bool nsScriptSecurityManager::sStrictFileOriginPolicy = true;
 
 namespace {
@@ -1371,22 +1370,28 @@
 
   mSystemPrincipal = system;
 
+  return NS_OK;
+}
+
+void nsScriptSecurityManager::InitJSCallbacks(JSContext* aCx) {
   //-- Register security check callback in the JS engine
   //   Currently this is used to control access to function.caller
-  sContext = danger::GetJSContext();
 
   static const JSSecurityCallbacks securityCallbacks = {
       ContentSecurityPolicyPermitsJSAction,
       JSPrincipalsSubsume,
   };
 
-  MOZ_ASSERT(!JS_GetSecurityCallbacks(sContext));
-  JS_SetSecurityCallbacks(sContext, &securityCallbacks);
-  JS_InitDestroyPrincipalsCallback(sContext, nsJSPrincipals::Destroy);
+  MOZ_ASSERT(!JS_GetSecurityCallbacks(aCx));
+  JS_SetSecurityCallbacks(aCx, &securityCallbacks);
+  JS_InitDestroyPrincipalsCallback(aCx, nsJSPrincipals::Destroy);
 
-  JS_SetTrustedPrincipals(sContext, system);
+  JS_SetTrustedPrincipals(aCx, BasePrincipal::Cast(mSystemPrincipal));
+}
 
-  return NS_OK;
+void nsScriptSecurityManager::ClearJSCallbacks(JSContext* aCx) {
+  JS_SetSecurityCallbacks(aCx, nullptr);
+  JS_SetTrustedPrincipals(aCx, nullptr);
 }
 
 static StaticRefPtr<nsScriptSecurityManager> gScriptSecMan;
@@ -1404,12 +1409,6 @@
 }
 
 void nsScriptSecurityManager::Shutdown() {
-  if (sContext) {
-    JS_SetSecurityCallbacks(sContext, nullptr);
-    JS_SetTrustedPrincipals(sContext, nullptr);
-    sContext = nullptr;
-  }
-
   NS_IF_RELEASE(sIOService);
   BundleHelper::Shutdown();
 }